### ISSUE TITLE
Feature request: add QuickNav command to jump to next/previous text paragraph

### ISSUE THREAD

**Issue by mltony** [nvaccess/nvda#15998]
I propose to add  QuickNav command to jump to next/previous text paragraph to NVDA. I I will implement this feature request myself if green light is given by NVDA team.

* Here we define text paragraph as paragraph with at least one sentence, that is written with punctuation marks, such as period, question mark, exclamation mark, their full-width CJK counterparts, etc.
* This allows users to quickly skip over clutter (such as menus, ads, etc) on a wide variety of websites.
* Another illustration of its utility is finding the beginning of an article on an unfamiliar news web site.
* This function is a heuristic that doesn't work 100% of the time. There are ads written in complete sentences, and likewise, sometimes in readable text of an article there can be a paragraph without any punctuation marks that will be missed. However, in practice this function works very well most of the time on a wide range of websites and simplifies navigation significantly.
* This feature has already been implemented in [TextNav add-on](https://addons.nvda-project.org/addons/textnav.en.html) and by default assigned to `Alt+Shift+Up/Down` gestures.
* The add-on is quite stable, it has been around for 5 years and for the last 4 years it's been in maintenance mode and didn't require any fixing besides NVDA compatibility updates.
* TextNav add-on is quite popular among NVDA users and has been frequently recommended on NVDA mailing list. Recently four users have conteacted me and asked to make it compatible with NVDA v2024.1. Moreover, a quick search reveals that people recommend TextNav in their blogs:
    * https://webfriendlyhelp.com/textnav/
    * (Misspelled as TestNav) https://slnconline.com.ng/10_nvda_addon_to_boost_your_productivity_and_effeciency.html
* In spirit this feature is similar to built-in N QuickNav command "Skip forward past a block of links", which in my understanding was also designed to skip menus on the websites. However in my experience jumping to text paragraph works much better then N QuickNav command as it works more reliably and typically finds exactly the first paragraph of the article.
* Finally, I believe this command will be quite useful for NVDA beginners, as one of the top tasks they struggle with is trying to make sense of web pages while trying to memorize dozens of keystrokes. With this command they'll be able to quickly and easily start using news articles, which I would think is one of the major use cases for the beginners.
* I propose to assign by default  QuickNav P and Shift+P gestures to this command.

**Comment by mltony** [comment]
Previous discussion in #15996.

**Comment by mltony** [comment]
@XLTechie, following our discussion in #15996 - then whoever created #12506 needs to clarify what does static text mean before we can even discuss that. These days most pages on the Internet are dynamic and typically partially generated by JavaScript, so I don't see a way to figure out programatically static text from this definition - as any text can be changed by JavaScript. And as for my definition - text without any formatting - again I don'tsee how this way it's going to be any more useful then text navigation proposed in this feature request.

**Comment by Adriani90** [comment]
@mltony thanks for splitting the features, it's easier to discuss.
I think the static text is something like this:
https://www.ibm.com/docs/en/spss-modeler/18.1.0?topic=controls-static-text

data:text/html,<StaticText text="static_text" textKey="text_key" > -- advanced custom layout options -- </StaticText>
However, checking only for this static text element is not enough.
So in my view the text paragrpah navigation should check following conditions:
1. The first word does not contain any role such as link, button, menu item etc.
2. Check for static text element
3. Check if it is a plain text and does not contain HTML (i.e. via c# by defining a regex). Here is an interesting discussion on this:
https://stackoverflow.com/questions/204646/how-to-validate-that-a-string-doesnt-contain-html-using-c-sharp

I still propose to enhance the n and shift+n command instead of having a new command due to following reasons:
1. The n and shift+n command (navigation by block of text before and after links) has been implemented with the exactly same use case in mind as requested in issue #71.
2. An attempt to achieve this has been implemented by @jcsteh in R3037 but I cannot find the coresponding commits or branches because it was probably on the old issue tracker.
3. User's expectation is to skip all the html roles and land on plain text when using this command, the same happens when pressing n or shift+n. So we would have kind of a double implementation and this will definitely cause confusions among users.

**Comment by mltony** [comment]
@Adriani90,

> I think the static text is something like this

That `<StaticText>` tag is not a part of standard HTML, this link points to product called IBM SPSS Modeler - I have no idea what's that and how that's relevant to this feature request and NVDA in general.

> 1.  The first word does not contain any role such as link, button, menu item etc

Meaningful paragraphs do contain links. They also do contain links in the first position. Check almost any wikipedia article. So checking for links is a wrong heuristic.

> 2. Check for static text element

Please find a valid definition of static text before we can argue about that.

> 3. Check if it is a plain text and does not contain HTML

Again, wrong heuristic. We would like to skip over Menus and Ads. None of them contain tags. In general rarely you would find a page on the Internet that contains tags, unless that's a page about web development.

> User's expectation is to skip all the html roles and land on plain text when using this command,

I can see that `N` command was probably conceived with roughly the same end goal in mind as paragraph navigation. However it's heuristic is wrong, as it is not very effective, at least in modern Web. Maybe it was more effective 15 years ago when it was first implemented, I don't remember what Internet looked like back then. 
1. Meaningful text contains links. Again think of Wiki for example.
2. Non-meaningful text sometimes doesn't contain links, so `N` lands you there.

Let's check how wel `N` command works in real world scenarios. Let me reuse British news website links from one of my previous comments:
1. https://www.bbc.com/news/world-africa-67826868
    * Pressing `N` once takes you to heading. Well maybe that's not unreasonable
    * Second `N` takes you to "Copyright 2024 BBC. "
    * Third `N` takes you to yet another notice "Beta Terms By using the Beta Site, "
    * No more text after a block of links. We missed the entire article. `N` is completely useless on this page.
2. https://www.theguardian.com/uk-news/2023/dec/29/tony-blair-rwanda-style-asylum-plan-2003
    * `N` finds 20 elements on the page, and none of them are actually inside the article. Again `N` is completely useless.
3. https://www.yorkpress.co.uk/news/24001621.yorks-electric-buses-just-cant-cope-winter-weather/
    * `N` finds 18 elements on the page
    * Finally, 1 out of 18 times it managed to land somewhere in the middle of article - it found a paragraph that starts with "The inward journey was cold "

Wow, I knew that `N` command was sort of useless, but this quick real world test shows that it is way worse and it is 99% useless. And these are just some random news links - not biased selection. Does anyone know any real NVDA user who uses `N` command ? Or maybe I am misunderstanding intended use of `N` command - in this case please let me know how it is supposed to be used.

Now back to the idea of merging paragraph navigation with `N` command. You are proposing to merge a highly effective and battle tested command (I hope I proved that enough) and a highly dysfunctional command. Why would anyone do that? The result is never going to be anything good.

Now I understand that having two commands would indeed confuse users. How about another idea:
* Let's call existing `N` command to be Paragraph navigation v1.
* Let's call this feature request to be Paragraph Navigation v2.
* Let's bind both of them to the same key - sure we can reuse `N`, and have a setting to let users choose between v1 and v2.
* Hopefully if you agree with the results of real world test above, you would agree that v2 should be the default option.

**Comment by Adriani90** [comment]
In this case why not replacing the current n quicknav command behavior with your proposed approach? If yours works better I don‘t see any reason why we should keep the old behavior. Just add „enhanced the behavior of finding plain text on webpages with n and shift+n“In changes.t2t file when implementing this and everyone will probably use more often this command afterwards.Maybe @jcsteh has other arguments on this because both approaches handle the same use case.Von meinem iPhone gesendetAm 03.01.2024 um 08:05 schrieb mltony ***@***.***>:﻿
@Adriani90,

I think the static text is something like this

That <StaticText> tag is not a part of standard HTML, this link points to product called IBM SPSS Modeler - I have no idea what's that and how that's relevant to this feature request and NVDA in general.


The first word does not contain any role such as link, button, menu item etc


Meaningful paragraphs do contain links. They also do contain links in the first position. Check almost any wikipedia article. So checking for links is a wrong heuristic.


Check for static text element


Please find a valid definition of static text before we can argue about that.


Check if it is a plain text and does not contain HTML


Again, wrong heuristic. We would like to skip over Menus and Ads. None of them contain tags. In general rarely you would find a page on the Internet that contains tags, unless that's a page about web development.

User's expectation is to skip all the html roles and land on plain text when using this command,

I can see that N command was probably conceived with roughly the same end goal in mind as paragraph navigation. However it's heuristic is wrong, as it is not very effective, at least in modern Web. Maybe it was more effective 15 years ago when it was first implemented, I don't remember what Internet looked like back then.

Meaningful text contains links. Again think of Wiki for example.
Non-meaningful text sometimes doesn't contain links, so N lands you there.

Let's check how wel N command works in real world scenarios. Let me reuse British news website links from one of my previous comments:

https://www.bbc.com/news/world-africa-67826868

Pressing N once takes you to heading. Well maybe that's not unreasonable
Second N takes you to "Copyright 2024 BBC. "
Third N takes you to yet another notice "Beta Terms By using the Beta Site, "
No more text after a block of links. We missed the entire article. N is completely useless on this page.


https://www.theguardian.com/uk-news/2023/dec/29/tony-blair-rwanda-style-asylum-plan-2003

N finds 20 elements on the page, and none of them are actually inside the article. Again N is completely useless.


https://www.yorkpress.co.uk/news/24001621.yorks-electric-buses-just-cant-cope-winter-weather/

N finds 18 elements on the page
Finally, 1 out of 18 times it managed to land somewhere in the middle of article - it found a paragraph that starts with "The inward journey was cold "



Wow, I knew that N command was sort of useless, but this quick real world test shows that it is way worse and it is 99% useless. And these are just some random news links - not biased selection. Does anyone know any real NVDA user who uses N command ? Or maybe I am misunderstanding intended use of N command - in this case please let me know how it is supposed to be used.
Now back to the idea of merging paragraph navigation with N command. You are proposing to merge a highly effective and battle tested command (I hope I proved that enough) and a highly dysfunctional command. Why would anyone do that? The result is never going to be anything good.
Now I understand that having two commands would indeed confuse users. How about another idea:

Let's call existing N command to be Paragraph navigation v1.
Let's call this feature request to be Paragraph Navigation v2.
Let's bind both of them to the same key - sure we can reuse N, and have a setting to let users choose between v1 and v2.
Hopefully if you agree with the results of real world test above, you would agree that v2 should be the default option.

**Comment by Adriani90** [comment]
Re static text: never mind I thoughht that would be HTML standard. But we should not focus on that anyway. I found here something like a definition for static content but I doubt this is being used broadly and it might be over complicated to identify such content programatically:
https://cloudinary.com/glossary/static-content

No idea what Voiceover on Mac defines as static text, but it is definitely nothing really official.

**Comment by mltony** [comment]
I just sent an email to NVDA mailing list titled "Does anyone use N QuickNav command in browse mode?" - to see if there are any users.
I always hesitate to mess with existing logic - who knows if anyone still using it and if we replace current command, then people might complain.
How about this: unless we receive many replies that people do still use `N` command, I will replace current `N` command logic with TextNav logic. Then if there are any complaints about `N` command gone missing - then I would introduce a setting to allow users to choose between v1 and v2. Does this sound good?

**Comment by mltony** [comment]
Ok, within minutes one user on NVDA mailing list responded that they like `N` command. I think based on this feedback we should preserve legacy functionality and go ahead with v1/v2 setting.

**Comment by nidza07** [comment]
Hello,
I personally very much agree that we should have a command to jump to next paragraph, but I don't understand the connection with the N command, and why we should mess with it at all, especially while the P key is completely free.

The N command has a very different purpose. Jump past a block of links doesn't necessarily mean that I want to jump to the start of an article, or even to any text at all. I cannot think of any examples off the top of my head right now, but in the past, I have used the N command on Websites where, for example, we have a menu with several links, and after that menu, some clickable elements for next/previous page. In that case, it's enough to focus the first link, and if we definitely know we aren't interested in anything on the first page, press N to skip that section entirely. The heuristic defined above wouldn't count such a scenario as a paragraph, and it even shouldn't.
As you say yourself, the feature is designed to skip past a menu, but skipping past a menu doesn't necessarily mean that I want to read an article.

So, my proposal is definitely implement the feature, but why not the P key as a paragraph? As I see that was even mentioned in the previous discussion, so apologies if an obvious answer was already given. Feel free to point me to it as I didn't have the time to read the entire discussion.

**Comment by Adriani90** [comment]
@mltony wrote:
> I just sent an email to NVDA mailing list titled "Does anyone use N QuickNav command in browse mode?" - to see if there are any users.

Thanks for that, this is a great idea. Let's see what users respond and you can certainly base your decision on the broader community opinion.

@nidza07 wrote:
> The N command has a very different purpose.

that's not entirely true. In #71 the user requested this feature to reach plain text on the website. What you describe is your own use case which you made out of the current implementation. However, the current approach is not modern anymore. Websites at that time didn't have that many elements with aria atributes, so links were the main elements with roles that users needed to skip.
Now adays we are dealing with a lot of interactive elements. Your use case is also covered when using d to jump to other landmarks or coma to skip the current container. Links that are next to each other are often within a container, so skipping the container with coma will most of the time do mainly the same job as the n command.

**Comment by brunoprietog** [comment]
It would be great to have this natively in NVDA. I use the TextNav add-on all the time and it certainly makes life more efficient.

**Comment by seanbudd** [comment]
Given the somewhat limited utility of this feature, it should perhaps remain an unassigned gesture, at least to begin with,

**Comment by mltony** [comment]
The utility of TextNav add-on (text paragraph navigation) is enormous. I would agree to leave all other gestures unassigned, but not this one. So with all respect, I stronglly urge you to reconsider. Let me reiterate my arguments.

* It can be used on virtually all webpages - and used meaningfully. Let me emphasize a few concrete  example use cases:
    * Browsing news websites - please see [this comment](https://github.com/nvaccess/nvda/discussions/15996#discussioncomment-7989404) where I illustrate that on 3 random British news websites it can find the beginning of the article in 1 or two keystrokes. I don't know whetehr you read news online, but many peoiple do, and I would think that people would find this extremely useful, because to my knowledge there is no other way to find the beginning of the article so quickly, and that works universally on most websites. If you can come up with alternative way to find the article on these three pages - please do let me know.
    * In general many other kinds of websites have a headers and navigation menu at the top, so TextNav allows you to skip them in a single keystroke.
    * Browsing forums. Typically most forum pages represent posts containing text that are separated by significant amount of clutter. I don't know how often you visit new forums but for my work I often need to do technical investigations and that involves googling "how to do X" and often times the results would lead mme to forums that I've never been to and if you don't want to spend the next 10 or 20 minutes learning the layout of that forum that you're not likely to visit again, then text paragraph navigation comes in emmensefly immensely handy.
    * Skipping ads. Many online ads contain multiple paragraphs, often more than 10. I don't know whetehr you run into this problem, but I am pretty sure that many NVDA users do. And with text paragraph navigation you can skip over an ad in just one keystroke, as opposed to 5 or 10 times pressing `Control+Down`.
* In my experience TextNav opens new use cases that otherwise would have been too tedious:
    * I can now read news aggregators, such as reddit or Hacker News, that contains links to news stories posted by other users that lead to other websites. Without TextNav it would have been too tedious to learn the layout of a new website every time. So TextNav closes the gap between NVDA users and sighted users.
    *( I can do technical investigations more easily as googling an error message typically might return results that lead to different forums. Sure one can learn stackoverflow.com, but what if the answer to your question has been asked on another platform.
* A couple of years ago I've been involved in teaching how to use computer to a couple of older people who recently lost their eyesight. In my experience they all immensely like TextNav. because it just works on almost every page. Without TextNav they would need to learn two dozens of other quickNav and arrow commands to reach similar level of proficiency. But if I only teach them TextNav, then they can get a rough idea of almost any website. This makes learning curve much gentler - as blind users can now start browsing some new websites with much less keystrokes to memorize.
* If we leave this gesture to be unassigned, this will essentially lock older users out of this feature, as for new users assigning shortcuts to a command would be as hard as installing an add-on. During my experience teaching older people, installing TextNav add-on was a blocker for nearly 100% new users (hopefully these days with add-on store things it became easier), but that's the reason why I would like to contribute this to NVDA core (hopefully with a default gesture) is to allow new users to learn browsing Internet more easily.

Also I looked at other QuickNav commands. `N` seems to me useless as I mentioned in [this comment](https://github.com/nvaccess/nvda/issues/15998#issuecomment-1874931794) - please correct me if I don't understand utility of `N` command. As for some other QuickNav commands, such as `A`, `S`, `W` - I personally haven't found a single website where they can be used - would be interesting to see if many users use them. And finally, we currently have unassigned letters: 'Z', 'Y', 'J', 'P' - and that's not counting punctuation marks that I'd think also can be used for QuickNav - what are we saving these for?

**Comment by michaelDCurran** [comment]
@mltony  thanks for your further info on this.
I would support p for paragraph (p and shift+p).

**Comment by simky316** [comment]
> I propose to add QuickNav command to jump to next/previous text paragraph to NVDA. I I will implement this feature request myself if green light is given by NVDA team.
> 
> * Here we define text paragraph as paragraph with at least one sentence, that is written with punctuation marks, such as period, question mark, exclamation mark, their full-width CJK counterparts, etc.
> * This allows users to quickly skip over clutter (such as menus, ads, etc) on a wide variety of websites.
> * Another illustration of its utility is finding the beginning of an article on an unfamiliar news web site.
> * This function is a heuristic that doesn't work 100% of the time. There are ads written in complete sentences, and likewise, sometimes in readable text of an article there can be a paragraph without any punctuation marks that will be missed. However, in practice this function works very well most of the time on a wide range of websites and simplifies navigation significantly.
> * This feature has already been implemented in [TextNav add-on](https://addons.nvda-project.org/addons/textnav.en.html) and by default assigned to `Alt+Shift+Up/Down` gestures.
> * The add-on is quite stable, it has been around for 5 years and for the last 4 years it's been in maintenance mode and didn't require any fixing besides NVDA compatibility updates.
> * TextNav add-on is quite popular among NVDA users and has been frequently recommended on NVDA mailing list. Recently four users have conteacted me and asked to make it compatible with NVDA v2024.1. Moreover, a quick search reveals that people recommend TextNav in their blogs:
>   
>   * https://webfriendlyhelp.com/textnav/
>   * (Misspelled as TestNav) https://slnconline.com.ng/10_nvda_addon_to_boost_your_productivity_and_effeciency.html
> * In spirit this feature is similar to built-in N QuickNav command "Skip forward past a block of links", which in my understanding was also designed to skip menus on the websites. However in my experience jumping to text paragraph works much better then N QuickNav command as it works more reliably and typically finds exactly the first paragraph of the article.
> * Finally, I believe this command will be quite useful for NVDA beginners, as one of the top tasks they struggle with is trying to make sense of web pages while trying to memorize dozens of keystrokes. With this command they'll be able to quickly and easily start using news articles, which I would think is one of the major use cases for the beginners.
> * I propose to assign by default  QuickNav P and Shift+P gestures to this command.

Thanks Tony for the add-on. I have been using it since you offered it several years ago. I use the add-on to navigate website, especially news sites and it saves me much time as I manage to skip other HTML elements, eg. buttons, links, etc. I tried the quick nav N and Shift + N but it does ot work well. It skips many blocks of text. I support incorporating it into NVDA core. Thanks.

**Comment by jacob-kruger-work** [comment]
I use it on a regular basis, and, am quite happy with default Alt+Shift+Up/Down keystrokes

**Comment by XLTechie** [comment]
@mltony I agree with your statements about utility; this feature is definitely 
not "of somewhat limited utility".
When I first started using NVDA years ago, it is the first thing I can remember 
saying "What? They don't do that?" about. :)

Answering your last question: a, s, w: the only one I find useful regularly, is 
s. Groups.io, for example, makes heavy use of separators.

**Comment by akash07k** [comment]
@mltony  I too use TextNav addon quite a lot. it is prooven extremely useful and I'll be very happy to see this functionality natively in NVDA

**Comment by surfer0627** [comment]
I use TextNav add-on everyday.
I' hope to see this feature "jump to next/previous text paragraph" natively in NVDA.

Q: Why is this feature necessary?

The existing QuickNav command in NVDA is undoubtedly valuable for navigation, but there's room for improvement. "Jumping to text paragraph works much better then N QuickNav command as it works more reliably and typically finds exactly the first paragraph of the article."

**Comment by Adriani90** [comment]
I think there is no doubt that users find this feature really useful, including myself.
@mltony from the discussion about n and shift+n, I have the feeling that the community has very diverse opinnions on this and it seems there is no common understanding of what it currently does. I suggest to remove that n and shift+n navigation feature completely and migrate people to the p and shift+p command which can be explained much better also in line with modern web design and serves the same use case but it provides a better solution.

**Comment by fisher729** [comment]
Hi.

In the user guide, it mentions jumping, but why not just say moving to be consistent?
New Quick Navigation command p for jumping to next/previous text paragraph in browse mode.
could be
New Quick Navigation command p for moving to the next/previous text paragraph in browse mode.

**Comment by fisher729** [comment]
And,
New unassigned Quick Navigation commands, which can be used to jump to the next/previous:
Could be:
New unassigned Quick Navigation commands, which can be used to move to the next/previous:

**Comment by seanbudd** [comment]
@fisher729 - this issue was closed 4 months ago and the changes for 2024.2 has been frozen for the final 2024.2 release on monday. we don't retroactively update our changes file after releases, so this will not be addressed in future. If you'd like to propose user guide changes please open a new issue/PR
